server/src/main/java/org/opensearch/index/query/QueryShapeVisitor.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.query;
+
+import org.apache.lucene.search.BooleanClause;
+import org.opensearch.common.SetOnce;
+
+import java.util.ArrayList;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Class to traverse the QueryBuilder tree and capture the query shape
+ */
+public class QueryShapeVisitor implements QueryBuilderVisitor {
+    private final SetOnce<String> queryType = new SetOnce<>();
+    private final Map<BooleanClause.Occur, List<QueryShapeVisitor>> childVisitors = new EnumMap<>(BooleanClause.Occur.class);
+
+    @Override
+    public void accept(QueryBuilder qb) {
+        queryType.set(qb.getName());
+    }
+
+    @Override
+    public QueryBuilderVisitor getChildVisitor(BooleanClause.Occur occur) {
+        // Should get called once per Occur value
+        if (childVisitors.containsKey(occur)) {
+            throw new IllegalStateException("child visitor already called for " + occur);
+        }
+        final List<QueryShapeVisitor> childVisitorList = new ArrayList<>();
+        QueryBuilderVisitor childVisitorWrapper = new QueryBuilderVisitor() {
+            QueryShapeVisitor currentChild;
+
+            @Override
+            public void accept(QueryBuilder qb) {
+                currentChild = new QueryShapeVisitor();
+                childVisitorList.add(currentChild);
+                currentChild.accept(qb);
+            }
+
+            @Override
+            public QueryBuilderVisitor getChildVisitor(BooleanClause.Occur occur) {
+                return currentChild.getChildVisitor(occur);
+            }
+        };
+        childVisitors.put(occur, childVisitorList);
+        return childVisitorWrapper;
+    }
+
+    String toJson() {
```

Comment:
Where are we using this method? just for testing?

Reply:
Using this only for testing

Reply:
Can we move to test class then?

Reply:
Discussed offline keeping as is for now